### PR TITLE
[Serve] Fix adding initial version and cleanup service versions

### PR DESCRIPTION
**Pull request by cblmemo**
Originally, in our serve state,
* We create the initial insertion to `version_specs` in the `__init__` function of `ReplicaManager`, which is not attached to insertion to the `services` table. This could sometimes lead to a misalign of the two tables;
* We do not cleanup the version information after its termination. Although we clean all outdated version in our controller process, this will left one final valid version at termination.

This PR fixed the two problems.

TODO: test.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: sky/serve/serve_utils.py [hunk 1]
```diff
@@ -414,21 +415,27 @@ def terminate_services(service_names: Optional[List[str]], purge: bool) -> str:
...
                                              with_replica_info=False)
-        assert service_status is not None, service_name
-        if service_status['status'] == serve_state.ServiceStatus.SHUTTING_DOWN:
+        if (service_status is not None and service_status['status']
+                == serve_state.ServiceStatus.SHUTTING_DOWN):
             # Already scheduled to be terminated.
             continue
-        if (service_status['status']
+        # If the `services` and `version_specs` table are not aligned, it might
+        # result in a None service status. In this case, the controller process
```
Comment (Michaelvll) [inline sky/serve/serve_utils.py:423]: Can we write down in what case the two tables will be misaligned if we have always handled the two database at the same time?

File: sky/serve/service.py [hunk 2]
```diff
@@ -156,6 +156,10 @@ def _start(service_name: str, tmp_task_yaml: str, job_id: int):
         with ux_utils.print_exception_no_traceback():
             raise ValueError(f'Service {service_name} already exists.')
 
+    # Add initial version information to the service state.
+    serve_state.add_or_update_version(service_name, constants.INITIAL_VERSION,
+                                      service_spec)
```
Comment (Michaelvll) [inline sky/serve/service.py:161]: Just curious, is this movement related to the bug?

**Review by Michaelvll — COMMENTED** [review]
Thanks for the fix @cblmemo! Left two quesetions.

File: sky/serve/serve_utils.py (hunk 1 above)
Comment (cblmemo) [reply sky/serve/serve_utils.py:423]: Previously, the system added an initial version in the `__init__`` function of replica managers and added the service record here https://github.com/skypilot-org/skypilot/blob/9e27a81eaee897dc0c8b61a44a65d7b543eb527c/sky/serve/service.py#L146-L151

If the code is aborted before adding the version and after adding the service, it will cause a misalignment. In this PR we move the adding of initial version closer to the service table, to reduce the possibility of causing such misalignment, though arguably it is still 'possible' to happed, so I added a safe guard here.

File: sky/serve/service.py (hunk 2 above)
Comment (cblmemo) [reply sky/serve/service.py:161]: Yes, please see the comment above 🫡

File: sky/serve/serve_utils.py (hunk 1 above)
Comment (cblmemo) [reply sky/serve/serve_utils.py:423]: An example of abortion would be `shutil` lib call to fail, e.g. permission issues

Comment (Michaelvll) [reply sky/serve/serve_utils.py:423]: Was the previous issue mainly because the clean up not deleting the final version from the `version_spec` table, which is fixed by the `delete_all_versions` now?

If this safe guard is only for rare case, it would be great to mention that this is rare in the comment here to make sure we still know whether this is a common case in the future. : )

Comment (cblmemo) [reply sky/serve/serve_utils.py:423]: Oh, sorry for not making it clear. The previous issue is the misalignment of `service` and `version_spec`, which causes the `_get_service_status` to return a `None` (due to the joining of the two tables), and causes the assertion to abort the deletion of the service.

https://github.com/skypilot-org/skypilot/blob/9e27a81eaee897dc0c8b61a44a65d7b543eb527c/sky/serve/serve_utils.py#L415-L417

I found that we did not clean up the remaining version when I debugged the original issue, and fix that in the same PR along the way.

Comment (cblmemo) [reply sky/serve/serve_utils.py:423]: > If this safe guard is only for rare case, it would be great to mention that this is rare in the comment here to make sure we still know whether this is a common case in the future. : )

Good point! Added a comment for this. Thanks!

**Review by Michaelvll — APPROVED** [review]
Thanks for the fix @cblmemo! LGTM.
